File: widget/customtextbrowser.py
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QPushButton, QTextBrowser


class CustomTextBrowser(QTextBrowser):
    def __init__(self, parent=None):
        super(CustomTextBrowser, self).__init__(parent)
        self.copyButton = QPushButton(self)
        self.copyButton.setStyleSheet(
            """
            QPushButton {
                image: url(:/copy.svg);
                background-color: transparent;
                border: none;
            }
            QPushButton:hover {
                background-color: rgba(115, 210, 22, 0.5);
                border: 1px solid rgb(115, 210, 22);
                border-radius: 5px;
            }
            QPushButton:pressed {
                background-color: rgba(80, 160, 17, 0.7);
                border: 1px solid rgb(80, 160, 17);
                border-radius: 5px;
            }
        """
        )
        self.copyButton.setCursor(Qt.PointingHandCursor)
        self.copyButton.clicked.connect(self.copyText)
        self.copyButton.hide()  # Initially hide the button
        # Set scrollbar width
        self.setStyleSheet(
            """
            QScrollBar:vertical {
                width: 10px;
            }
        """
        )

    def enterEvent(self, event):
        self.adjustButtonPosition()
        self.copyButton.show()
        super(CustomTextBrowser, self).enterEvent(event)

    def leaveEvent(self, event):
        self.copyButton.hide()
        super(CustomTextBrowser, self).leaveEvent(event)

    def resizeEvent(self, event):
        self.adjustButtonPosition()
        super(CustomTextBrowser, self).resizeEvent(event)

    def adjustButtonPosition(self):
        scrollbar_width = 10 if self.verticalScrollBar().isVisible() else 0
        self.copyButton.move(
            self.width() - self.copyButton.width() - scrollbar_width, 0
        )

    def copyText(self):
        self.selectAll()
        self.copy()
        self.moveCursor(self.textCursor().End)

File: src/widget/customtextbrowser.cpp
//
// Created by pzy123 on 4/15/2024.
//
#include "customtextbrowser.h"

CustomTextBrowser::CustomTextBrowser(QWidget *parent) : QTextBrowser(parent) {
    setStyleSheet("QScrollBar:vertical { width: 15px; }");
    copyButton = new QPushButton(this);
    copyButton->setStyleSheet("QPushButton {"
                              "image: url(:/copy);"
                              "}"
                              "QPushButton:hover {"
                              "background-color: rgb(115,210,22);"
                              "}");
    copyButton->setFixedSize(QSize(25, 25));
    connect(copyButton, &QPushButton::clicked, this, &CustomTextBrowser::copyText);
    copyButton->hide(); // Initially hide the button
//    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
}

void CustomTextBrowser::adjustCopyButtonPosition() {
    int margin = 5; // Margin from the edge
    int buttonXPosition;

    QScrollBar *vScrollBar = this->verticalScrollBar();
    if (vScrollBar->isVisible()) {
        // If the vertical scroll bar is visible, position the button taking into account the scroll bar's width
        buttonXPosition = this->width() - copyButton->width() - vScrollBar->width() - margin;
    } else {
        // If the vertical scroll bar is not visible, position the button at the edge
        buttonXPosition = this->width() - copyButton->width() - margin;
    }

    // Adjust the button's position. Assuming the button's Y position is fixed or determined elsewhere
    copyButton->move(buttonXPosition, copyButton->y());
}
void CustomTextBrowser::enterEvent(QEvent *event) {
    QTextBrowser::enterEvent(event);
    adjustCopyButtonPosition();
    copyButton->show();
}

void CustomTextBrowser::leaveEvent(QEvent *event) {
    QTextBrowser::leaveEvent(event);
    copyButton->hide();
}

void CustomTextBrowser::resizeEvent(QResizeEvent *event) {
    QTextBrowser::resizeEvent(event);
    adjustCopyButtonPosition();
}

void CustomTextBrowser::copyText() {
    selectAll();
    copy();
    moveCursor(QTextCursor::End);
}
